src/Features/Core/Portable/MoveToNamespace/AbstractMoveToNamespaceService.cs

```diff
@@ -0,0 +1,139 @@
+        {
+            var typeAnalysisResult = await AnalyzeTypeAtPositionAsync(document, span.Start, cancellationToken).ConfigureAwait(false);
+
+            if (typeAnalysisResult.CanPerform)
+            {
+                return ImmutableArray.Create(AbstractMoveToNamespaceCodeAction.Generate(this, typeAnalysisResult));
+            }
+
+            return ImmutableArray<AbstractMoveToNamespaceCodeAction>.Empty;
+        }
+
+        internal override async Task<MoveToNamespaceAnalysisResult> AnalyzeTypeAtPositionAsync(
+            Document document,
+            int position,
+            CancellationToken cancellationToken)
+        {
+#if DEBUG // TODO: remove once the feature is done
+            var root = await document.GetSyntaxRootAsync().ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var token = root.FindToken(position);
+            var node = token.Parent;
+
+            var symbolInfo = semanticModel.GetSymbolInfo(node, cancellationToken: cancellationToken);
+            var symbol = symbolInfo.Symbol;
+            string @namespace = null;
+
+            if (symbol is INamespaceSymbol namespaceSymbol && !(node is TNamespaceDeclarationSyntax))
+            {
+                node = node.FirstAncestorOrSelf<SyntaxNode>(a => a is TNamespaceDeclarationSyntax);
+            }
+
+            if (node is TNamespaceDeclarationSyntax declarationSyntax)
+            {
+                if (ContainsNamespaceDeclaration(node))
+                {
+                    return new MoveToNamespaceAnalysisResult("Namespace container contains nested namespace declaration");
+                }
+
+                @namespace = GetNamespaceName(declarationSyntax);
+                return new MoveToNamespaceAnalysisResult(document, node, @namespace, MoveToNamespaceAnalysisResult.ContainerType.Namespace);
+            }
+
+            if (symbol is INamedTypeSymbol namedTypeSymbol)
+            {
+                node = node.FirstAncestorOrSelf<SyntaxNode>(a => a is TNamedTypeDeclarationSyntax);
+                @namespace = GetQualifiedName(namedTypeSymbol.ContainingNamespace);
+            }
+
+            if (node is TNamedTypeDeclarationSyntax namedTypeDeclarationSyntax)
+            {
+                @namespace = @namespace ?? GetNamespaceName(namedTypeDeclarationSyntax);
+                return new MoveToNamespaceAnalysisResult(document, node, @namespace, MoveToNamespaceAnalysisResult.ContainerType.NamedType);
+            }
+
+            return new MoveToNamespaceAnalysisResult("Not a valid position");
+#else 
+
+            return await Task.FromResult(new MoveToNamespaceAnalysisResult("Feature is not complete yet")).ConfigureAwait(false);
+#endif
+        }
+
+        private bool ContainsNamespaceDeclaration(SyntaxNode node)
+            => node.DescendantNodes(n => n is TCompilationSyntax || n is TNamespaceDeclarationSyntax)
```

Comment:
What's up with the `TCompilationSyntax` here?

Reply:
Leftover code that I should have removed :)